src/dioptra/restapi/v1/jobs/service.py

```diff
@@ -527,6 +525,157 @@ def get(
+        metrics = [
+            {"name": metric, "value": run.data.metrics[metric]}
+            for metric in run.data.metrics.keys()
+        ]
+
+        return metrics
+
+    def update(
+        self,
+        job_id: int,
+        metric_name: str,
+        metric_value: float,
+        **kwargs,
+    ) -> dict[str, Any]:
+        """Update a job's metrics by its unique id.
+
+        Args:
+            job_id: The unique id of the job.
+            metric_name: The name of the metric to create or update.
+            metric_value: The value of the metric being updated.
+        Returns:
+            The metric dictionary passed in if successful, otherwise an error message.
+        """
+        log: BoundLogger = kwargs.get("log", LOGGER.new())
+        log.debug("Update job metrics by id", job_id=job_id)
+
+        from mlflow.tracking import MlflowClient
+
+        run_id: UUID = self._job_id_mlflowrun_service.get(job_id=job_id, **kwargs)[
+            "mlflow_run_id"
+        ]
+        client = MlflowClient()
+        client.log_metric(run_id.hex, key=metric_name, value=metric_value)
+
+        # this is here just to raise an error if the run does not exist
+        try:
+            run = client.get_run(run_id.hex)  # noqa: F841
+        except MlflowException as e:
+            raise EntityDoesNotExistError("MlFlowRun", run_id=run_id.hex) from e
+
+        return {"name": metric_name, "value": metric_value}
+
+
+class JobIdMetricsSnapshotsService(object):
+    """The service methods for retrieving the historical metrics of a
+    job by unique id and metric name."""
+
+    @inject
+    def __init__(self, job_id_mlflowrun_service: JobIdMlflowrunService) -> None:
+        """Initialize the job metrics snapshots service.
+
+        All arguments are provided via dependency injection.
+        """
+        self._job_id_mlflowrun_service = job_id_mlflowrun_service
+
+    def get(
+        self,
+        job_id: int,
+        metric_name: str,
+        page_index: int,
+        page_length: int,
+        **kwargs,
+    ) -> tuple[list[dict[str, Any]], int]:
+        """Fetch a job's metrics by its unique id and metric name.
+
+        Args:
+            job_id: The unique id of the job.
+            metric_name: The name of the metric.
+            page_index: The index of the first page to be returned.
+            page_length: The maximum number of experiments to be returned.
+        Returns:
+            The metric history for the requested job and metric if found,
+            otherwise an error message.
+        """
+        from mlflow.tracking import MlflowClient
+
+        log: BoundLogger = kwargs.get("log", LOGGER.new())
+        log.debug(
+            "Get job metric history by id and name",
+            job_id=job_id,
+            metric_name=metric_name,
+        )
+
+        run_id: UUID = self._job_id_mlflowrun_service.get(job_id=job_id, **kwargs)[
+            "mlflow_run_id"
+        ]
+        client = MlflowClient()
```

Comment:
If it can't connect to the mlflow server it attempts to re-connect several times, so this request can hang for a bit. Not sure if this is the best behavior? Maybe we should set number of retries lower?